src/lib.rs

```diff
@@ -1472,28 +1473,35 @@ pub mod __rt {
         }
 
         pub fn borrow(&self) -> Ref<T> {
-            unsafe {
-                if self.borrow.get() == usize::max_value() {
-                    borrow_fail();
-                }
-                self.borrow.set(self.borrow.get() + 1);
-                Ref {
-                    value: &*self.value.get(),
-                    borrow: &self.borrow,
-                }
+            match self
+                .borrow
+                .fetch_update(Ordering::Acquire, Ordering::Relaxed, |borrow| {
+                    (borrow != usize::max_value()).then(|| borrow + 1)
+                }) {
+                Ok(_) => unsafe {
+                    Ref {
+                        value: &*self.value.get(),
+                        borrow: &self.borrow,
+                    }
+                },
+                Err(_) => borrow_fail(),
             }
```

Comment:
Continuation from https://github.com/rustwasm/wasm-bindgen/pull/3610#discussion_r1325773792:
the body of match will never be executed unless `fetch_update()` succeeds, this can't be reordered. So the question here is does this need to be coordinated with another access? As far as I can see: no.

E.g. dropping `RefMut` can happen before, in which case this succeeds, or happen after, in which case it doesn't succeed. Either outcome is satisfactory to us and doesn't allow unsafe behavior.

Reply:
> So the question here is does this need to be coordinated with another access? As far as I can see: no.

It does: it needs to be coordinated with any potential writes that another thread did while holding a `RefMut`. At the very least, the destructor for `RefMut` needs to use `Release` ordering, since otherwise the compiler (and processor) is free to reorder writes to `self.value` after setting `self.borrow` back to 0. And then that needs to match up with an `Acquire` here.

`Ref`'s destructor also still needs `Release` ordering because the compiler's free to reorder reads from `self.value` to happen after `self.borrow` gets set back to 0, by which point another thread might have already started modifying the value.

> the body of this `match` will never be executed unless `fetch_update()` succeeds, this can't be reordered.

You're probably right that the compiler can't reorder it, but I think it's still possible for subsequent reads from `self.value` to read outdated values anyway, e.g. if that value's cache line is still in the CPU's [invalidation queue](https://en.wikipedia.org/wiki/MESI_protocol#Invalidate_Queues).

I wouldn't be surprised if CPUs do all kinds of other weird stuff that I don't know about: I only found out about invalidation queues by skimming https://marabos.nl/atomics because this seemed sketchy. So I think we should try and make this work according to the theoretical memory model, rather than honing in too much on exactly what the compiler's doing.

Reply:
> At the very least, the destructor for `RefMut` needs to use `Release` ordering, since otherwise the compiler (and processor) is free to reorder writes to `self.value` after setting `self.borrow` back to 0.

> `Ref`'s destructor also still needs `Release` ordering because the compiler's free to reorder reads from `self.value` to happen after `self.borrow` gets set back to 0, by which point another thread might have already started modifying the value.

Why would reading/writing to `self.value` be able to be reordered at all? What am I missing here?

> So I think we should try and make this work according to the theoretical memory model, rather than honing in too much on exactly what the compiler's doing.

I agree, but I definitely don't know enough about atomic operations, maybe I should read that whole book, thanks for sharing!

Reply:
https://doc.rust-lang.org/1.72.0/std/cell/struct.UnsafeCell.html:
> At all times, you must avoid data races. If multiple threads have access to the same `UnsafeCell`, then any writes must have a proper happens-before relation to all other accesses (or use atomics).

Is what I was looking for I guess.